maps-compose/src/main/java/com/google/maps/android/compose/MapViewLifecycleController.kt

```diff
@@ -0,0 +1,136 @@
+package com.google.maps.android.compose
+
+import android.os.Bundle
+import android.util.Log
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.Event
+import androidx.lifecycle.LifecycleEventObserver
+import com.google.android.gms.maps.MapView
+import com.google.maps.android.compose.MapViewLifecycleController.LifecycleDirection.Down
+import com.google.maps.android.compose.MapViewLifecycleController.LifecycleDirection.Up
+
+internal class MapViewLifecycleController(
+    isMapReused: Boolean,
+    private val mapView: MapView
+) {
+    private val lcTag = run {
+        val mapViewId = mapView.tagData().debugId
+        "MVLC/$mapViewId"
+    }
+
+    private companion object {
+        val lifecycleUp = listOf(
+            Event.ON_CREATE,
+            Event.ON_START,
+            Event.ON_RESUME
+        )
+    }
+
+    private enum class LifecycleDirection {
+        Up, Down
+    }
+
+    private var previousLifecycleState = if (isMapReused)
+        Event.ON_STOP else
+        Event.ON_CREATE
+
+    var lifecycle: Lifecycle? = null
+        set(value) {
+            if (field !== value) {
+                field?.removeObserver(observer)
+                value?.addObserver(observer)
+                field = value
+            }
+        }
+
+    /**
+     * Move to the lifecycle event instead of setting it directly.
+     * */
+    private fun moveToLifecycleEvent(targetLifecycle: Event) {
+        if (targetLifecycle == previousLifecycleState) return
+
+        val lifecycleDirection = if (targetLifecycle in lifecycleUp) Up else Down
+
+        (if (lifecycleDirection == Up) "UP" else "DOWN").let { strDirection ->
+            Log.d(lcTag, "Moving $strDirection from $previousLifecycleState to $targetLifecycle.")
+        }
+
+        do {
+            val nextLifecycleState = nextLifecycleEvent(lifecycleDirection)
+            setLifecycleEvent(nextLifecycleState)
+        } while (nextLifecycleState != targetLifecycle)
+    }
+
+    private fun nextLifecycleEvent(direction: LifecycleDirection) = when (previousLifecycleState) {
+        Event.ON_CREATE -> when (direction) {
+            Up -> Event.ON_START
+            Down -> error("No lifecycle event below ON_CREATE.")
+        }
+
+        Event.ON_START -> when (direction) {
+            Up -> Event.ON_RESUME
+            Down -> error("No lifecycle event below ON_START.")
+        }
+
+        Event.ON_RESUME -> when (direction) {
+            Up -> error("No lifecycle event above ON_RESUME.")
+            Down -> Event.ON_PAUSE
+        }
+
+        Event.ON_PAUSE -> when (direction) {
+            Up -> Event.ON_RESUME
+            Down -> Event.ON_STOP
+        }
+
+        Event.ON_STOP -> when (direction) {
+            Up -> Event.ON_START
+            Down -> Event.ON_DESTROY
+        }
+
+        Event.ON_DESTROY -> error("No lifecycle event above ON_DESTROY")
+        Event.ON_ANY -> error("Unsupported operation")
```

Comment:
Maybe make this just `else` like below? I find it confusing to see a case of `ON_ANY`.

Reply:
Will do